### PR TITLE
any resolution

### PR DESCRIPTION
**Pull request by rfomin**
* Use modified `R_DrawColumn` for patches scaling. Taken from Eternity Engine.

* Restore old `SCREENWIDTH` and `SCREENHEIGHT` macros, use `video_t` structure instead.

Alternative to #1292 

This is very much incomplete. Currently it will set the internal resolution to 1536x720 if `hires` is enabled. `hires` has been removed from most places.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Über-impressive! You just rewrote the while f*cking renderer! 😮

**Comment by rfomin** [comment]
> Über-impressive! You just rewrote the while f*cking renderer! 😮

No, just part of the `v_video.c` 😄 The original renderer already works at any resolution.

**Comment by ceski-1** [comment]
This is great! Seems like a good starting point to build from. I'm curious to see what it takes to eliminate the hardcoded parts like Eternity Engine and Rum and Raisin did. And of course, multithreading would be a killer feature to help address the performance scaling issues with traditional software renderers.

**Comment by rfomin** [comment]
> And of course, multithreading would be a killer feature to help address the performance scaling issues with traditional software renderers.

I want to try multithreading, but that would require a lot of work. In the next release I think we can implement dynamic resolution. I plan to have 3 options in the resolution menu: original, double and high, the last one will be dynamic.

**Comment by rfomin** [comment]
This is mostly done, I think this PR is getting too big.

What is missing:
* "Blocky" fuzz
* FPS dots (I think we can remove them)

TODO:
* Dynamic resolution
* Change resolution on window resize
* Menu resolution slider or some other UI

**Comment by MrAlaux** [comment]
> * Menu resolution slider or some other UI

Why not just input the resolution as a `[Width]x[Height]` string, like in Eternity? Of course, automatically rounding it to the closest multiple of 4.

**Comment by rfomin** [comment]
> Why not just input the resolution as a `[Width]x[Height]` string, like in Eternity? Of course, automatically rounding it to the closest multiple of 4 once entered.

Eternity's resolution UI is very bad. Even I don't know what numbers to enter.

**Comment by fabiangreffrath** [comment]
> * "Blocky" fuzz

I think we need this, *especially* with increasing resolution.

> * FPS dots (I think we can remove them)

Fine with me, bye.

**Comment by fabiangreffrath** [comment]
> > * Menu resolution slider or some other UI
> 
> Why not just input the resolution as a `[Width]x[Height]` string, like in Eternity? Of course, automatically rounding it to the closest multiple of 4 once entered.

Because that's a prime example of bad UI that puts all burden on the player and demands them to know exactly what should be entered. I'd rather expect a limited choice of "blocky as in Vanilla Doom", "intermediate as used to from Woof", "higher than that".

**Comment by MrAlaux** [comment]
I'm aware that this is still in progress, but just in case, I want to report some weird artifacting in the rightmost column of my 1366x768 screen:

woof0000

Also, walls appear darker in this PR compared to the last release (both with High Resolution off), but flats seem to be unaffected:

ezgif com-add-text

File: src/r_segs.c [hunk 1]
```diff
@@ -157,7 +157,7 @@ void R_RenderMaskedSegRange(drawseg_t *ds, int x1, int x2)
       {
         if (!fixedcolormap)      // calculate lighting
           {                             // killough 11/98:
-            unsigned index = spryscale>>(LIGHTSCALESHIFT+hires);
+            unsigned index = spryscale >> (LIGHTSCALESHIFT + (video.xscale >> FRACBITS));
```
Comment (rfomin) [inline src/r_segs.c:160]: > Also, walls appear darker in this PR compared to the last release (both with High Resolution off), but flats seem to be unaffected:

This is one of the places that I'm not sure about.

**Comment by rfomin** [comment]
> I'm aware that this is still in progress, but just in case, I want to report some weird artifacting in the rightmost column of my 1366x768 screen

Thanks for feedback, I'll look into it.

**Comment by rfomin** [comment]
> > * "Blocky" fuzz
> 
> I think we need this, _especially_ with increasing resolution.

Sure, I'd appreciate the help 😉

File: src/r_segs.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/r_segs.c:160]: Try to divide `spryscale` by the vertical scaling factor prior to `>> LIGHTSCALESHIFT`.

https://github.com/fabiangreffrath/woof/pull/1292/files#diff-4fcc8d4da7773102015f5441efb6d009a3d1fb277a6463c751f4bf199109d7c0R160

**Comment by rfomin** [comment]
> I want to report some weird artifacting in the rightmost column of my 1366x768 screen

@MrAlaux https://github.com/fabiangreffrath/woof/pull/1298/commits/f1c7295b6d6aa64e18c2f25e7b03a5d4defa63f8 fixed it for me.

**Comment by fabiangreffrath** [comment]
> Sure, I'd appreciate the help 😉

Ouch! 😬

File: src/r_segs.c (hunk 1 above)
Comment (rfomin) [reply src/r_segs.c]: Thanks, this helps: https://github.com/fabiangreffrath/woof/pull/1298/commits/6fbc2a1594e4dc3d880ceb4079a70c5c2b233a1b Still not perfect I think.

Comment (fabiangreffrath) [reply src/r_segs.c]: It will be perfect for integer scale. Maybe use fixed point multiplication instead?

**Comment by fabiangreffrath** [comment]
Any chance we get the `-1`, `-2`, `-3` command line parameters back to working?

**Comment by rfomin** [comment]
> Any chance we get the `-1`, `-2`, `-3` command line parameters back to working?

I want to try changing the resolution when the window is resized and dynamic resolution. To do this we want to change the resolution as fast as possible, so major changes to `i_video.c` may be needed. The question is when to stop, split into multiple PRs or keep working with this?

**Comment by liPillON** [comment]
Sorry to bother you all mid-dev, buut..


> Currently it will set the internal resolution to 1536x720 if `hires` is enabled.

is this still true for the latest commits/builds?



> I'd rather expect a limited choice of "blocky as in Vanilla Doom", "intermediate as used to from Woof", "higher than that".

If this is still on the table, I would like to suggest having more than 3 options.
Having to chose only between "Vanilla/LoRes", "MBF/HiRes" and "Native/Auto", would mean that people disliking how 400p looks would choose "Native" and this might not be ideal for systems with slow CPUs hooked up to big panels (think low-power HTPCs used also for light gaming on 4K TVs)

**Comment by fabiangreffrath** [comment]
> is this still true for the latest commits/builds?

No, the latest state is to max out resolution based on the screen's dimensions.

> If this is still on the table, I would like to suggest having more than 3 options. Having to chose only between "Vanilla/LoRes", "MBF/HiRes" and "Native/Auto", would mean that people disliking how 400p looks would choose "Native" and this might not be ideal for systems with slow CPUs hooked up to big panels (think low-power HTPCs used also for light gaming on 4K TVs)

I've thought about this as well. There should be something between 640x480 and Full-HD or even 4K.

**Comment by fabiangreffrath** [comment]
> Edit: Alternatively, I can edit the commit history so we don't squash it when merging. 

Yes, we shouldn't squash this, else it introduced too many important code changes at once.

**Comment by fabiangreffrath** [comment]
> The question is when to stop, split into multiple PRs or keep working with this?

I'm fine with merging this branch as soon as the major regressions are fixed. To me, this is currently only the missing blocky fuzz rendering. But I have an idea for this, so if you grant me some time, I'll try sto see where I can get.

**Comment by rfomin** [comment]
> Having to chose only between "Vanilla/LoRes", "MBF/HiRes" and "Native/Auto", would mean that people disliking how 400p looks would choose "Native" and this might not be ideal for systems with slow CPUs hooked up to big panels (think low-power HTPCs used also for light gaming on 4K TVs)

The idea of Auto Mode is to dynamically adjust the resolution on every frame, like most modern games. It will always target for 60 FPS and drop the resolution to 600p minimum. We will see if it works.

**Comment by liPillON** [comment]
My proposal:
- save the WxH resolution in pixels to the config file, a-la dsda (so power users can tinker with it)
- give users a limited set of choices in the UI (that the engine would translate to the correct WxH size and save to the cfg, after having taken widescreen and aspect correction into account)

As for the options, I would use a terminology familiar to both power users and casual gamers:
- LoRes (200/240p)
- HiRes (400/480p)
- HD	(600/720p)
- FullHD(900/1080p)
- QHD	(1200/1440p)
- UHD/4K(1800/2160p)

**Comment by fabiangreffrath** [comment]
> * Auto	(Native)
> * LoRes (200/240p)
> * HiRes (400/480p)
> * HD	(600/720p)
> * FullHD (900/1080p)

I'd even stop here. We are talking about the (30 years old) software renderer after all.

> * QHD (1200/1440p)
> * UHD/4K (1800/2160p)

**Comment by rfomin** [comment]
> ah neat, so it's "auto" as in "dinamic" and not as in "automatically detected"

Yes, correct.

**Comment by rfomin** [comment]
> To me, this is currently only the missing blocky fuzz rendering. But I have an idea for this, so if you grant me some time, I'll try sto see where I can get.

It seems to work in your branch, we can adapt it at least. I'll try to do something too.

**Comment by fabiangreffrath** [comment]
> It seems to work in your branch, we can adapt it at least. I'll try to do something too.

Yes, but we have integer scaling factors there. What I think needs to be done here:
* Draw only every `n`th column with `n = V_ScaleX(1)`
* Convert `count` to fixed point
* Iterate over `count`. During each step:
  * `memset(dest, fuzz, V_ScaleX(1))`
  * `dest += linewidth`
  * `count -= video.ystep`
  * If `count` crossed a FRACUNIT boundary, recalculate `fuzz`.

File: src/r_segs.c (hunk 1 above)
Comment (rfomin) [reply src/r_segs.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/1298/commits/4243e3a2e2846d5539b032b9c1f0fac2a799902a

**Comment by fabiangreffrath** [comment]
But what if `video.yscale = 2.5 * FRACUNIT`?

**Comment by fabiangreffrath** [comment]
I think it needs `nx * linesize` in the `fuzz` calculations.

**Comment by rfomin** [comment]
> I think it needs `nx * linesize` in the `fuzz` calculations.

This almost fixed it.

> The test wad https://github.com/fabiangreffrath/woof/pull/1217 is a good way to check if fuzz is working correctly.

Wow, this test file is beautiful, thanks!

**Comment by fabiangreffrath** [comment]
> This almost fixed it.

Wait, it skips lines not rows, so this should be `ny`.

**Comment by rfomin** [comment]
It still looks a bit different from vanilla and has some artifacts:
image

I guess we have to fix it the hard way.

**Comment by fabiangreffrath** [comment]
FWIW, I could live with small glitches during the split second of a resolution change. Regarding the bottom line artifacts, I guess this needs a `>=` condition in the `cutoff` check.

**Comment by fabiangreffrath** [comment]
Also, it should match very well for integer scaling factors. Visual glitches should only occur when switching from or to oddball resolutions - and I think we can live with that.

**Comment by fabiangreffrath** [comment]
> So we will have to choose "special" multipliers or maybe just integers for this feature.

That'd be fine with me anyway.

**Comment by rfomin** [comment]
> I remember when we first discussed about DRS and you mentioned that it would default to 600p, or did I misunderstand that?

Now I think 600p should be the minimum. By the way, does this work on Linux?

**Comment by fabiangreffrath** [comment]
Yes, it works under Linux, though I only experienced one resolution switch so far.

**Comment by rfomin** [comment]
> Yes, it works under Linux, though I only experienced one resolution switch so far.

Try Eviternity map26. Also you can tweak parameters, for example target 500 FPS here:

`if (time < 1000000ull / 60) // 60 FPS  target`

Also try to lower minimum resolution here (e.g. set 240):
```
if (render_height < 720)
{
    render_height = 720;
}
```
You can get fast resolution changes, every 5 ms. I'm curious about the stability on other systems.

**Comment by liPillON** [comment]
hey how is it going?
I too wanted to try Eviternity MAP26 using the latest build.

After spawning I've moved to the railing in front of the player and pointed the crosshair towards one of the two towers in the distance (or any area between them).
This has always been a spot where performance dipped with every port I've tried.

Here, the hud vanished; the sky started to wobble horizontally, just like what happens when switching the linear sky setting on/off; the edges of each "polygon" kept on "vibrating".
I got the impression that the engine was stuck switching resolutions every other frame and this was taking priority over any other drawing routine.

**Comment by rfomin** [comment]
> Here, the hud vanished; the sky started to wobble horizontally, just like what happens when switching the linear sky setting on/off; the edges of each "polygon" kept on "vibrating".
> I got the impression that the engine was stuck switching resolutions every other frame and this was taking priority over any other drawing routine.

You are using the Boom HUD, right? Yes, HUD widgets disappear for some reason, but other things should work. "Vibrating" is the resolution changing every 5 ms, it's due to precision errors of Doom's fixed math. In the future we will choose parameters so that it does not vibrate. Currently the settings are pushed to the extreme for testing purposes. Does it run well? No crashes/locks etc? Try it with classic status bar :smile:

**Comment by rfomin** [comment]
> 1366x768 display

It's not 1368x768? I didn't know such displays existed. What resolution is SDL reporting? Try running `woof -verbose`, it will print the resolution to the console.

> At other angles, it keeps switching and doesn't settle on a resolution at all.

The idea was to make these switches seamless, but the scaling is not accurate enough. It seems we have to be very careful when choosing resolution multipliers.

Anyway, I'm glad that SDL is stable and fast enough at surface/texture recreating, I was thinking about rewriting `i_video.c` with OpenGL. Thanks for testing!

**Comment by MrAlaux** [comment]
> Are you absolutely sure it's 1366x768?

Windows itself doesn't let me go beyond 1366x768. Many, many games cap my resolution at that value too.

As far as I knew, 1366x768 is not only a real resolution, but also a rather common one. According to a little research, it's WXGA.
See:
https://en.wikipedia.org/wiki/Display_resolution#Common_display_resolutions
https://en.wikipedia.org/wiki/Graphics_display_resolution#1366x768

**Comment by rfomin** [comment]
> As far as I knew, 1366x768 is not only a real resolution, but also a rather common one. According to a little research, it's WXGA.

Oh, I only saw 1368x768 displays for some reason. Okay, I think we can calculate the aspect ratio more accurately, it will probably fix the back borders.

**Comment by fabiangreffrath** [comment]
> It still looks a bit different from vanilla and has some artifacts

This ad-hoc patch fixes that for me:

```patch
--- a/src/r_draw.c
+++ b/src/r_draw.c
@@ -542,9 +542,11 @@ static void R_DrawFuzzColumn_block(void)
   if (!dc_yl)
     dc_yl = ny;
 
-  if (dc_yh == viewheight - ny)
+  // [FG] shift whole column so count remains a multiple of ny
+  while (dc_yh >= viewheight - ny - 1)
   {
-    dc_yh = viewheight - 2 * ny;
+    dc_yl--;
+    dc_yh--;
     cutoff = true;
   }
 
```

**Comment by rfomin** [comment]
> This ad-hoc patch fixes that for me:

Hmm, it crashes for me. What resolution are you using?

**Comment by fabiangreffrath** [comment]
That was some odd-ball resolution to force overdrawing of the status bar with fuzz columns. Something like 405p or so.

**Comment by rfomin** [comment]
Oh sorry, it was unrelated issue. Looks good for me.

**Comment by rfomin** [comment]
> This ad-hoc patch fixes that for me:

On some resolutions it is now one pixel too high (200p):
image

**Comment by fabiangreffrath** [comment]
Yay integer rounding!

**Comment by fabiangreffrath** [comment]
And while at it, may I suggest to decrease resolution only is steps of 100p, so we don't end up with the most odd-ball resolutions which require quirks on their own. Also, I'd limit resolutions between 400p and 1000p in dynamic mode, 400p still counts as "high". 😉

**Comment by rfomin** [comment]
@fabiangreffrath I'm working on it. I just realized that it's not necessary to recreate surfaces, SDL can copy rectangles. So we can allocate all arrays/buffers on the native resolution and then just change the logical size, which makes resolution switching much faster.

I also want to try to calculate the DRS factor more accurately, which might reduce the "vibration".

> Also, I'd limit resolutions between 400p and 1000p in dynamic mode, 400p still counts as "high".

I don't think we should go below 600p. We can just keep the 400p option in the menu. We'll see when it's ready.

> Also, I think this needs some kind of a tolerance interval for the target refresh rate, i.e. only act if below target rate minus 10% and above target rate plus 10%, for damping out short peak values.

Good idea.

**Comment by fabiangreffrath** [comment]
> @fabiangreffrath I'm working on it. 

I know. I didn't mean to push you, just share some shower thoughts. 😁 


> I just realized that it's not necessary to recreate surfaces, SDL can copy rectangles. So we can allocate all arrays/buffers on the native resolution and then just change the logical size, which makes resolution switching much faster.

Awesome!

> I don't think we should go below 600p. We can just keep the 400p option in the menu. We'll see when it's ready.

Alright, fine with me. 👍 for the menu entry.

File: src/r_main.c [hunk 2]
```diff
@@ -558,7 +558,7 @@ void R_ExecuteSetViewSize (void)
         }
     }
 
-    HU_disable_all_widgets();
+    //HU_disable_all_widgets();
```
Comment (rfomin) [inline src/r_main.c:561]: @fabiangreffrath Why do we need this line?

**Comment by liPillON** [comment]
tried the latest build: it loads up, intializes the music and then crashes without error messages
this happens regardless of both the fullscreen and hires settings used
the last two lines in my verbose log file:
`I_ResetResolution: 2136x1201`
`S_ChangeMusic: D_DM2TTL (DOOM2.WAD)`

I have a 16:10 1920x1200 display, though..

**Comment by fabiangreffrath** [comment]
It crashes in `V_FillRect()`.

**Comment by rfomin** [comment]
> `I_ResetResolution: 2136x1201` `S_ChangeMusic: D_DM2TTL (DOOM2.WAD)`
> 
> I have a 16:10 1920x1200 display, though..

Oops, that's a buffer overflow. I was tweaking for a 16:9 aspect ratio.

File: src/i_video.c [hunk 3]
```diff
@@ -420,28 +433,62 @@ static inline void I_UpdateRender (void)
     }
 }
 
+#define DRS_MIN_HEIGHT 400
+#define DRS_STEP       1
+#define DRS_MIN_DELTA  10
+static int drs_percent = 100;
+static boolean drs_update;
```
Comment (rfomin) [inline src/i_video.c:440]: @fabiangreffrath  With the latest changes, the "vibration" has been significantly reduced. I'm not sure if it can be removed completely, it appears even with integer multipliers. But I think that with a smarter approach to these parameters we can make it almost imperceptible.

Comment (fabiangreffrath) [reply src/i_video.c:440]: This is good news! I expect resolution changes to happen not too often during casual play, so a little visual shake won't be much of a problem.

File: src/r_main.c (hunk 2 above)
Comment (fabiangreffrath) [reply src/r_main.c:561]: I don't remember anymore. Had something to do with dynamic hud element arrangement.

**Comment by fabiangreffrath** [comment]
This happens with the latest build:


https://github.com/fabiangreffrath/woof/assets/6449237/6ba2accf-d2f3-45fd-a410-39a159cbac23

**Comment by rfomin** [comment]
> This happens with the latest build:

Yes, it's far from perfect, but it was much worse before. Not sure how to fix it yet. We can make DRS more conservative, but then it's not much use because on some maps the complexity is drastically different depending on the viewpoint.

**Comment by fabiangreffrath** [comment]
I'd like to push this effort over the finish line. So, what's left to do?

Dymanic resolution works in general, but is not perfect. There are sports on certain maps where the frame rate drops below the threshold, triggering a resolution reduction, which in turn leads to the rendering framerate rising over the threshold again. But this is with the player standing still. Maybe we can turn this into a condition that the resolution isn't changed twice in the same spot?

Apart from this, we still need a means to explicitly select at least a few common resolutions in the menu. At least 200p, 400p, 600p and "native" should be available e.g. by means of a slider with "dynamic" being the last option after "native".

And then, that's it, right?

**Comment by rfomin** [comment]
> Dymanic resolution works in general, but is not perfect.

DRS is not good in its current state, both in terms of code and usability. I am currently in the process of refactoring it.
Alternatively, we can drop DRS commits for now and do it in a separate PR.

> And then, that's it, right?

Fuzz also not good.

**Comment by fabiangreffrath** [comment]
> DRS is not good in its current state, both in terms of code and usability. I am currently in the process of refactoring it. Alternatively, we can drop DRS commits for now and do it in a separate PR.

That'd be fine with me. The faster we have any_res code in the master branch, the better.

> Fuzz also not good.

But this only affects the cutoff line at the bottom, right?

**Comment by rfomin** [comment]
> The faster we have any_res code in the master branch, the better.

Ok, I'll try to finish DRS this evening, and if I fail, I'll drop it and prepare a commit history.

> > Fuzz also not good.
> 
> But this only affects the cutoff line at the bottom, right?

I'm not sure, but that's the most noticeable issue. I don't think it's good enough in its current state.

File: src/i_video.c [hunk 4]
```diff
@@ -416,32 +472,65 @@ static inline void I_UpdateRender (void)
...
+    if (actualpercent > DRS_GREATER)
+    {
+        double reduction = (actualpercent - DRS_GREATER ) * 0.2;
+        newheight = (int)MAX(DRS_MIN_HEIGHT, oldheight - oldheight * reduction);
+    }
+    else if (actualpercent < DRS_LESS)
+    {
+        double addition = (DRS_LESS - actualpercent) * 0.25;
+        newheight = (int)MIN(renderer_height, oldheight + oldheight * addition);
+    }
```
Comment (rfomin) [inline src/i_video.c:509]: @fabiangreffrath, @mikeday0 I tried to implement your previous suggestions, it works better but still not perfect. Corrections, or other ideas?

Comment (fabiangreffrath) [reply src/i_video.c:508]: Much better! Though, sometimes the resolution switches back to 1916x900, whereas my original resolution would be 1920x900. Also, there are some odd-ball resolutions in between: 1840x864 1500x704 1136x536 1672x784. Is there any system behind this?

File: src/d_loop.c [hunk 5]
```diff
@@ -728,7 +728,7 @@ void TryRunTics (void)
 
     // [AM] If we've uncapped the framerate and there are no tics
     //      to run, return early instead of waiting around.
-    #define return_early (uncapped && counts == 0 && leveltime > oldleveltime)
+    #define return_early (uncapped && counts == 0)
```
Comment (ceski-1) [inline src/d_loop.c:731]: This code is correct, but just to avoid confusion, it's not always 60 fps in the menus. Removing `leveltime > oldleveltime` simply allows the menus to run at the same framerate as the rest of the game. That's what I wanted, so that's okay. Between `uncapped`, `fpslimit`, `use_vsync`, and the video driver settings, there could be a bunch of different scenarios but they're all fine to me.

File: src/i_video.c (hunk 4 above)
Comment (rfomin) [reply src/i_video.c:508]: > Though, sometimes the resolution switches back to 1916x900, whereas my original resolution would be 1920x900. 

Currently we only set the height and the width is calculated in `ResetResolution`. I can try to calculate the aspect ratio more precisely.

> Also, there are some odd-ball resolutions in between: 1840x864 1500x704 1136x536 1672x784. Is there any system behind this?

It is trying to optimize the FPS, see the `DynamicResolution` function. The question is whether these resolutions look bad or not, the exact number doesn't matter. Perhaps we should not calculate the height directly, but choose from an array of predefined resolutions.

Comment (fabiangreffrath) [reply src/i_video.c:508]: > Currently we only set the height and the width is calculated in `ResetResolution`. I can try to calculate the aspect ratio more precisely.

Hm, if `height = 900` is passed it should always result in the same width, or not?

> Perhaps we should not calculate the height directly, but choose from an array of predefined resolutions.

I think I'd prefer that. It would feel less "alien". 😉

Comment (fabiangreffrath) [reply src/i_video.c:508]: > Hm, if `height = 900` is passed it should always result in the same width, or not?

I think rounding up the `height` value to the next multiple of `4` should happen first in that function.

Comment (fabiangreffrath) [reply src/i_video.c:508]: > I think I'd prefer that. It would feel less "alien". 😉

Maybe only allow steps of, say, 40px.

Comment (rfomin) [reply src/i_video.c:508]: > Hm, if `height = 900` is passed it should always result in the same width, or not?

There's probably a bug somewhere. I plan to calculate everything using `float`/`double`.

**Comment by rfomin** [comment]
I give up with DRS at the moment, it works but not good enough for the default resolution mode. Currently it's the default mode, because I think we'll make resolution UI after updating the menus in another PR. The rest is done.

File: src/i_video.c (hunk 4 above)
Comment (mikeday0) [reply src/i_video.c:514]: This works really well on my system. Nice work, I am super impressed! So far I've been testing with Eviternity Map 26. In the opening area I can get it to switch resolutions by switching between looking at the tower complex and the less-detailed area to the left. The resolution switching is very snappy, and I haven't seen any instances of instability.

Comment (rfomin) [reply src/i_video.c:514]: Thanks, but I suspect you're using pretty powerful hardware and the resolution doesn't drop too low? Currently `DRS_MIN_HEIGHT` is 400p, and there is a  "vibration" noticeable on status bar or weapons rendering when switching resolutions quickly. I'm thinking of setting the minimum to 600p, but then the DRS feature wouldn't be very useful for 1080p displays.

Comment (mikeday0) [reply src/i_video.c:514]: i7-4770K, 16 GB of RAM, GTX 770. Good hardware 10 years ago. :)  My display is 1920 x 1200, 60Hz. The lowest resolution I see when running around in Eviternity map 26 is 1152 x 720. (Reported with `-verbose` console output, not the IDRATE cheat.)

Comment (rfomin) [reply src/i_video.c:514]: I guess we need an even more extreme map for testing 😄 I tested it on an old 2011 Core i5 laptop, it drops to a minimum 853x480 (or 400p without aspect ration correction). With a powerful CPU at high resolutions, the "vibration" is much less noticeable.

**Comment by rfomin** [comment]
@fabiangreffrath I have edited the commit history, ready for review.

File: src/i_video.c (hunk 4 above)
Comment (MrAlaux) [reply src/i_video.c:515]: > I guess we need an even more extreme map for testing

Map31 of Antaresian Reliquary might be useful. Maybe Eviternity II has heavier maps, since it is a new release.

**Comment by rfomin** [comment]
> I'm sorry to hear that and I understand your frustration. For me, it's working good enough now, but I agree that it may need more fine-tuning before it may become the new default.

Oh, I probably chose the wrong tone. It annoys me that this PR gets too big and lags behind the master, causing merge conflicts. I'm definitely going to tweak the DRS some more after the merge.

> I'm afraid that it may take some time until the restructuring of the menu is finished

I plan to get something done this weekend. I've been waiting for this since 2021 https://github.com/fabiangreffrath/woof/issues/110 :wink:

Anyway, I will add resolutions to the menu.

**Comment by fabiangreffrath** [comment]
> Oh, I probably chose the wrong tone. It annoys me that this PR gets too big and lags behind the master, causing merge conflicts. I'm definitely going to tweak the DRS some more after the merge.

Please feel free to merge this at your convenience. I have already reviewed most parts of the code step by step over the development of this branch, and what I've seen was awesome. There is now way I could review this PR in its entirety again.

> I plan to get something done this weekend. I've been waiting for this since 2021 #110 😉

Seems like pragmatism has replaced ideology - and shows that I 'm still able to learn. 😁 

> Anyway, I will add resolutions to the menu.

Great, thank you!

**Review by fabiangreffrath — APPROVED** [review]

File: src/m_misc.c [hunk 6]
```diff
@@ -134,9 +134,9 @@ default_t defaults[] = {
   //
 
   { // killough 11/98: hires
-    "hires", (config_t *) &default_hires, NULL,
-    {1}, {0,1}, number, ss_none, wad_no,
-    "1 to enable 640x400 resolution for rendering scenes"
+    "resolution_mode", (config_t *) &default_resolution_mode, NULL,
+    {RES_DRS}, {RES_ORIGINAL, NUM_RES - 1}, number, ss_none, wad_no,
+    "0 - original 200p, 1 - double 400p, 2 - triple 300p, 4 - native (dynamic)"
```
Comment (liPillON) [inline src/m_misc.c:139]: typo: 300p instead of 600p :)

Comment (rfomin) [reply src/m_misc.c]: Fixed, thanks!
